tidy: enable bugprone-use-after-move

fanquake opened the pull request:
Would have caught #25640.

Currently `// NOLINT`s around:
```bash
test/util_tests.cpp:2513:34: error: 't2' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    BOOST_CHECK(v2[0].origin == &t2);
                                 ^
test/util_tests.cpp:2511:15: note: move occurred here
    auto v2 = Vector(std::move(t2));
              ^
test/util_tests.cpp:2519:34: error: 't2' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    BOOST_CHECK(v3[1].origin == &t2);
                                 ^
test/util_tests.cpp:2516:15: note: move occurred here
    auto v3 = Vector(t1, std::move(t2));
              ^
test/util_tests.cpp:2527:34: error: 't3' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    BOOST_CHECK(v4[2].origin == &t3);
                                 ^
test/util_tests.cpp:2523:15: note: move occurred here
    auto v4 = Vector(std::move(v3[0]), v3[1], std::move(t3));
```

See: https://releases.llvm.org/14.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/bugprone-use-after-move.html

fanquake committed:
tidy: enable bugprone-use-after-move

Will error with:
```bash
coins.cpp:102:22: error: 'coin' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
           (uint32_t)coin.nHeight,
                     ^
coins.cpp:96:21: note: move occurred here
    it->second.coin = std::move(coin);
```

until #25663 is merged.

See:
https://releases.llvm.org/14.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/bugprone-use-after-move.html

jonatack commented:
Concept ACK, looks like it is finding cases in `test/util_tests.cpp` as well.

ryanofsky commented:
Concept ACK, since this does seem like it could catch unintentional mistakes, but I have a little skepticism because it can be perfectly ok to use an object after moving. For example, vector's move constructor guarantees the moved from object will be empty (https://en.cppreference.com/w/cpp/container/vector/vector).

Currently some things are failing in CI as jonatack pointed out:

https://github.com/bitcoin/bitcoin/blob/e25b0c6838f375c810bee292ed49e13c9b5acc69/src/test/util_tests.cpp#L2495-L2497

But maybe we could work around these with a custom move function (`util::SafelyMoveFrom`?) that does the same as `std::move` but avoids triggering the error. Or, I wonder if something like `std::move(*&t2)` would trick the linter

fanquake committed:
test: work around bugprone-use-after-move warnings in util tests

```bash
test/util_tests.cpp:2513:34: error: 't2' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    BOOST_CHECK(v2[0].origin == &t2);
                                 ^
test/util_tests.cpp:2511:15: note: move occurred here
    auto v2 = Vector(std::move(t2));
              ^
test/util_tests.cpp:2519:34: error: 't2' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    BOOST_CHECK(v3[1].origin == &t2);
                                 ^
test/util_tests.cpp:2516:15: note: move occurred here
    auto v3 = Vector(t1, std::move(t2));
              ^
test/util_tests.cpp:2527:34: error: 't3' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    BOOST_CHECK(v4[2].origin == &t3);
                                 ^
test/util_tests.cpp:2523:15: note: move occurred here
    auto v4 = Vector(std::move(v3[0]), v3[1], std::move(t3));
```

fanquake commented:
> Or, I wonder if something like std::move(*&t2) would trick the linter

Looks like it does (locally). I've pushed this up for now, but will defer to you as to what change / how you'd rather handle this in the util tests.

ajtowns commented on src/test/util_tests.cpp:
```diff
@@ -2508,19 +2508,19 @@ BOOST_AUTO_TEST_CASE(test_tracked_vector)
-    auto v2 = Vector(std::move(t2));
+    auto v2 = Vector(std::move(*&t2));
     BOOST_CHECK_EQUAL(v2.size(), 1U);
     BOOST_CHECK(v2[0].origin == &t2);
```
Shouldn't you just write `// NOLINT(*-use-after-move)` to disable the check for these lines (which also documents that you're deliberately doing weird things) ? https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics

ryanofsky replied:
Makes sense, I agree `// NOLINT(*-use-after-move)` looks better than what I suggested (`SafelyMoveFrom(t2)` or `std::move(*t2)`). I didn't know what syntax to disable linter was, but I assumed it would be worse than that.

I think ideally the compiler or linter would have a usable-after-moved directive that could be applied to types like `std::vector` where using after move is safe and well defined, so individual moves wouldn't have to be annotated. But either way works

fanquake replied:
Switched to `// NOLINT`.

ryanofsky reviewed (approved):
Code review ACK 696434d7a03a4e028484142aa2d80e4f0a63c60c. It's nice to enable tidy checks to catch unintended uses after moves. Workaround for intended uses seems ok, and `NOLINT` alternate suggestion would be ok too.

ryanofsky reviewed (approved):
Code review ACK f345dc3960c2cf4d69ebbcc011e4e836205f0361. Only change since last review is switching to NOLINT directives
